File: Array/4.median_of_two_sorted_arrays.py
import sys
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        x = len(nums1)
        y = len(nums2)
        if x > y:
            return self.findMedianSortedArrays(nums2, nums1)
        low = 0
        high = x
        while low <= high:
            partition_x = (low+high)//2
            partition_y = (x+y+1)//2 - partition_x
            max_left_x = -sys.maxsize if partition_x == 0 else nums1[partition_x-1]
            min_right_x = sys.maxsize if partition_x == x else nums1[partition_x]
            max_left_y = -sys.maxsize if partition_y == 0 else nums2[partition_y-1]
            min_right_y = sys.maxsize if partition_y == y else nums2[partition_y]
            if max_left_x <= min_right_y and max_left_y <= min_right_x:
                if (x+y) % 2 == 0:
                    return (max(max_left_x,max_left_y) + min(min_right_x, min_right_y))/2
                else:
                    return float(max(max_left_x, max_left_y))
            elif max_left_x > min_right_y:
                high = partition_x -1
            else:
                low = partition_x + 1

File: C++/Binary Search/4. Median of Two Sorted Arrays.cpp
#include <algorithm>
#include <vector>

class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int totalLength = nums1.size() + nums2.size();
        if (totalLength % 2 == 1) {
            return getKthElement(nums1, nums2, (totalLength+1)/2);
        }
        else {
            return (getKthElement(nums1, nums2, totalLength/2) + getKthElement(nums1, nums2, totalLength/2+1)) / 2;
        }
    }
    
    double getKthElement(const vector<int>& nums1, const vector<int>& nums2, int k) {
        int index1 = 0, index2 = 0;
        int m = nums1.size(), n = nums2.size();
        int newIndex1, newIndex2;
        while (true) {
            // 特殊情况
            if (index1 == m) {
                return (double)nums2[index2 + k - 1];
            }
            if (index2 == n) {
                return (double)nums1[index1 + k - 1];
            }
            if (k == 1) {
                return (double)min(nums1[index1], nums2[index2]);
            }
            // 一般情况
            newIndex1 = min(index1 + k/2-1, m-1);
            newIndex2 = min(index2 + k/2-1, n-1);
            if (nums1[newIndex1] <= nums2[newIndex2]) {
                k -= newIndex1 - index1 + 1;
                index1 = newIndex1 + 1;
            }
            else {
                k -= newIndex2 - index2 + 1;
                index2 = newIndex2 + 1;
            }
        }
    }
};